[JSC] Use quadratic-probing in PropertyTable

Constellation opened the pull request:
#### f5a9b7f2385a5386bf94ab68ca3f2e8df3374f21
<pre>
[JSC] Use quadratic-probing in PropertyTable
<a href="https://bugs.webkit.org/show_bug.cgi?id=251899">https://bugs.webkit.org/show_bug.cgi?id=251899</a>
rdar://105156586

Reviewed by Keith Miller and Michael Saboff.

This patch makes PropertyTable better with two changes.

1. `reinsert` should not do key comparison. `reinsert` happens only when we insert already hash-consed keys. This
   means any keys should not be equal. This makes loop of `reinsert` significantly simpler.
2. This patch integrates quadratic-probing into our PropertyTable instead of linear-probing. This works much better
   for objects with many keys. This improved JetStream2/json-stringify-inspector by 2.24%. And JetStream2/string-unpack-code-SP
   also gets better by 2.07%.

* Source/JavaScriptCore/runtime/PropertyTable.h:
(JSC::PropertyTable::findImpl):
(JSC::PropertyTable::reinsert):

Canonical link: <a href="https://commits.webkit.org/260031@main">https://commits.webkit.org/260031@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b90aa373e72bcbb86964d0e9cab11d2fbf1c69c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106804 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39587 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115987 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/17320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7030 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98992 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112572 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/17320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/96113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/17320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/27765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/96334 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/9004 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/29120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/95800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6939 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9571 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/95800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/15173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48664 "Passed tests") | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/104556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11097 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/37/builds/104556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3739 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

msaboff reviewed (approved):
r=me

kmiller68 commented on Source/JavaScriptCore/runtime/PropertyTable.h:
```diff
@@ -532,13 +535,24 @@ inline void PropertyTable::reinsert(Index* indexVector, Entry* table, const Valu
+    // Reinsert must not conflict with the keys since all entries are existing ones.
+    // Plus, there is no deleted entries too. We should just check emptyness, that's it.
+    while (true) {
+        if (indexVector[index] == EmptyEntryIndex)
```
Can we add an assert that none of the keys are going to be deleted? IIUC, reinsert is only called when growing the property table.

Constellation replied:
Done

kmiller68 reviewed (approved):
r=me too with nit.

Constellation committed:
[JSC] Use quadratic-probing in PropertyTable
https://bugs.webkit.org/show_bug.cgi?id=251899
rdar://105156586

Reviewed by Keith Miller and Michael Saboff.

This patch makes PropertyTable better with two changes.

1. `reinsert` should not do key comparison. `reinsert` happens only when we insert already hash-consed keys. This
   means any keys should not be equal. This makes loop of `reinsert` significantly simpler.
2. This patch integrates quadratic-probing into our PropertyTable instead of linear-probing. This works much better
   for objects with many keys. This improved JetStream2/json-stringify-inspector by 2.24%. And JetStream2/string-unpack-code-SP
   also gets better by 2.07%.

* Source/JavaScriptCore/runtime/PropertyTable.h:
(JSC::PropertyTable::findImpl):
(JSC::PropertyTable::reinsert):

Canonical link: https://commits.webkit.org/260031@main
